lxd/api_project.go

```diff
@@ -1078,6 +1080,252 @@ func projectNodeConfigDelete(d *Daemon, s *state.State, name string) error {
+	// Get the project entities and config.
+	var usedBy []string
+	var projectConfig map[string]string
+	err := s.DB.Cluster.Transaction(ctx, func(ctx context.Context, tx *db.ClusterTx) error {
+		project, err := dbCluster.GetProject(ctx, tx.Tx(), projectName)
+		if err != nil {
+			return fmt.Errorf("Fetch project %q: %w", projectName, err)
+		}
+
+		usedBy, err = projectUsedBy(ctx, tx, project)
+		if err != nil {
+			return err
+		}
+
+		projectConfig, err = dbCluster.GetProjectConfig(ctx, tx.Tx(), project.Name)
+		if err != nil {
+			return fmt.Errorf("Fetch project %q config: %w", projectName, err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Parse and delete entities.
+	defaultProfile := api.NewURL().Path(version.APIVersion, "profiles", api.ProjectDefaultName).Project(projectName).String()
+	localNodeName := s.ServerName
+
+	for _, u := range usedBy {
+		if u == defaultProfile {
+			continue
+		}
+
+		url, err := url.Parse(u)
+		if err != nil {
+			return err
+		}
+
+		entityType, _, target, pathArgs, err := entity.ParseURL(*url)
+		if err != nil {
+			return err
+		}
+
+		// Skip entities that are not local to this node (only applies to instances and storage volumes).
+		if (entityType == entity.TypeInstance || entityType == entity.TypeStorageVolume) && target != "" && target != localNodeName {
+			continue
+		}
+
+		// Delete the entity based on its type.
+		switch entityType {
+		case entity.TypeInstance:
+			instanceName := pathArgs[0]
+			err = s.Authorizer.CheckPermission(ctx, entity.InstanceURL(projectName, instanceName), auth.EntitlementCanDelete)
+			if err != nil {
+				return err
+			}
+
+			// Load instance.
+			inst, err := instance.LoadByProjectAndName(s, projectName, instanceName)
+			if err != nil {
+				return err
+			}
+
+			if inst.State() != strings.ToUpper(api.Stopped.String()) {
+				// Stop instance.
+				err = doInstanceStatePut(inst, api.InstanceStatePut{
+					Action:  "stop",
+					Timeout: -1,
+					Force:   true,
+				})
+				if err != nil {
+					return err
+				}
+			}
+
+			// Delete instance.
+			op, err := doInstanceDelete(ctx, s, instanceName, projectName)
+			if err != nil {
+				return err
+			}
+
+			err = op.Start()
+			if err != nil {
+				return err
+			}
+
+			err = op.Wait(ctx)
+			if err != nil {
+				return err
+			}
+
+		case entity.TypeProfile:
+			profileName := pathArgs[0]
+			err = s.Authorizer.CheckPermission(ctx, entity.ProfileURL(projectName, profileName), auth.EntitlementCanDelete)
+			if err != nil {
+				return err
+			}
+
+			// Delete profile.
+			err = doProfileDelete(ctx, s, profileName, projectName)
+			if err != nil {
+				return err
+			}
+
+			// Delete default profile if "features.profiles" is enabled.
+			if shared.IsTrue(projectConfig["features.profiles"]) {
+				err = s.DB.Cluster.Transaction(ctx, func(ctx context.Context, tx *db.ClusterTx) error {
+					return dbCluster.DeleteProfile(ctx, tx.Tx(), projectName, "default")
```

Comment:
do we even need this part? profiles are just DB records linked to the project itself, and when the project record is removed the profile records will be too.